File: Class_06_Frontend_Interviews_And_Merge_Sort/Frontend_Interviews_and_Merge_Sort_Homework/Solutions/merge_sort_solution.py
def merge_sort(arr):
    merge_sort2(arr, 0, len(arr) - 1)

def merge_sort2(arr, first, last):
    if first < last:
        middle = (first + last)//2
        merge_sort2(arr, first, middle)
        merge_sort2(arr, middle+1, last)
        merge(arr, first, middle, last)

def merge(arr, first, middle, last):
    left = arr[first:middle+1]
    right = arr[middle+1:last+1]
    left.append(999999999)
    right.append(999999999)
    i = j = 0

    for k in range (first, last+1):
        if left[i] <= right[j]:
            arr[k] = left[i]
            i += 1
        else:
            arr[k] = right[j]
            j += 1

input1 = [45, 98, 3, 24, 15, 77, 9, 50] # [3, 9, 15, 24, 45, 50, 77, 98]
input2 = [18, 16, 27, 4, 12] # [4, 12, 16, 18, 27]

print("input1")
print(input1)
merge_sort(input1)
print(input1)

print("input2")
print(input2)
merge_sort(input2)
print(input2)

File: Unit_11_Sorting/Homework/Solutions/mergeSortSolution.js
// helper function: merges two sorted sub-arrays into one sorted array
function merge(arr1, arr2) {
    // create holding array to combine values
    let arrFinal = []
    let i = 0
    let j = 0

    // compare arrays element by element, add to arrFinal in order
    while(i < arr1.length && j < arr2.length) {
        if (arr2[j] >= arr1[i]) {
            arrFinal.push(arr1[i])
            i++
        }
        else {
            arrFinal.push(arr2[j])
            j++
        }
    }

    // after one array is out of elements
    // add the rest of the other array's elements to arrFinal (they're already sorted)
    while (i < arr1.length) {
        arrFinal.push(arr1[i])
        i++
    }
    while(j < arr2.length) {
        arrFinal.push(arr2[j])
        j++
    }

    return arrFinal
}

// main function
function mergeSort(arr) {
    // recursive base case
    if (arr.length <= 1) return arr;

    // divide the array in half (divide and conquer approach)
    // recursively split array halves into single elements
    let mid = Math.floor(arr.length / 2)
    let left = mergeSort(arr.slice(0, mid))
    let right = mergeSort(arr.slice(mid))

    return merge(left, right)
}

const input1 = [45, 98, 3, 24, 15, 77, 9, 50] // [3, 9, 15, 24, 45, 50, 77, 98]
const input2 = [18, 16, 27, 4, 12] // [4, 12, 16, 18, 27]

console.log(mergeSort(input1))
console.log(mergeSort(input2))